xmodule/item_bank_block.py

```diff
@@ -0,0 +1,503 @@
+
+    Concretely, this is the shared base for ItemBankBlock and LegacyLibraryContentBlock.
+    Sharing fields and selection logic between those two blocks will allow us to eventually migrate
+    LegacyLibraryContentBlocks (backed by V1 libraries) into ItemBankBlocks (backed by V2 libraries).
+    """
+    has_children = True
+    has_author_view = True
+    show_in_read_only_mode = True
+    resources_dir = 'assets/library_content'
+    mako_template = 'widgets/metadata-edit.html'
+    studio_js_module_name = "VerticalDescriptor"
+
+    max_count = Integer(
+        display_name=_("Count"),
+        help=_("Enter the number of components to display to each student. Set it to -1 to display all components."),
+        default=1,
+        scope=Scope.settings,
+    )
+    selected = List(
+        # This is a list of (block_type, block_id) tuples used to record
+        # which random/first set of matching blocks was selected per user
+        default=[],
+        scope=Scope.user_state,
+    )
+    # This cannot be called `show_reset_button`, because children blocks inherit this as a default value.
+    allow_resetting_children = Boolean(
+        display_name=_("Show Reset Button"),
+        help=_("Determines whether a 'Reset Problems' button is shown, so users may reset their answers and reshuffle "
+               "selected items."),
+        scope=Scope.settings,
+        default=False
+    )
+
+    @classproperty
+    def completion_mode(cls):  # pylint: disable=no-self-argument
+        """
+        Allow overriding the completion mode with a feature flag.
+
+        This is a property, so it can be dynamically overridden in tests, as it is not evaluated at runtime.
+        """
+        if settings.FEATURES.get('MARK_LIBRARY_CONTENT_BLOCK_COMPLETE_ON_VIEW', False):
+            return XBlockCompletionMode.COMPLETABLE
+
+        return XBlockCompletionMode.AGGREGATOR
+
+    @classmethod
+    def make_selection(cls, selected, children, max_count):
+        """
+        Dynamically selects block_ids indicating which of the possible children are displayed to the current user.
+
+        Arguments:
+            selected - list of (block_type, block_id) tuples assigned to this student
+            children - children of this block
+            max_count - number of components to display to each student
+
+        Returns:
+            A dict containing the following keys:
+
+            'selected' (set) of (block_type, block_id) tuples assigned to this student
+            'invalid' (set) of dropped (block_type, block_id) tuples that are no longer valid
+            'overlimit' (set) of dropped (block_type, block_id) tuples that were previously selected
+            'added' (set) of newly added (block_type, block_id) tuples
+        """
+        rand = random.Random()
+
+        selected_keys = {tuple(k) for k in selected}  # set of (block_type, block_id) tuples assigned to this student
```

Comment:
Oh right, this 😛

It's because we assign lists-of-tuples to `self.selected`, but when you save the block and then load it back from ModuleStore, `self.selected` comes back as list-of-2-element-lists. So we need to handle both cases here by converting to tuples. I'll add a comment.

(Back in the last iteration of this project, I had [straightened this out and made it type safe](https://github.com/openedx/edx-platform/pull/33511/files#diff-49b6eaf1d7d27a5c33e95e4f444a3c6306f49790419f6488679667f663171864R495-R507), but that never merged. I can't bite that off again right now, but I'd love to do it some time before the Sumac cut.)

Reply:
Left a comment